servoshell: Add `ParseResolutionError` to parse elegantly

Integral-Tech opened the pull request:
When passing an invalid resolution string (such as `1x1x1`) to the `--screen-size` or `--window-size` argument, Servo starts without any error. Additionally, if the width or height is set to 0, Servo crashes with a SIGSEGV (Address boundary error).

This patch addresses the following issue by several changes:

1. Introduce a custom error type ParseResolutionError.
2. Replace the `split()` method with `split_once()`.
3. Make the capital 'X' an acceptable separator.
4. Add a check to prevent crashes when width or height is set to 0.

---

Before: 
```
╰─❯ ./servo --screen-size=0
index out of bounds: the len is 1 but the index is 1 (thread main, at ports/servoshell/prefs.rs:236)
fish: Job 1, './servo --screen-size=0' terminated by signal SIGSEGV (Address boundary error)
```
```
╰─❯ ./servo --screen-size=0x1
xdg_surface#30: error -1: invalid window geometry size (0x1)
assertion `left != right` failed
  left: 0
 right: 0 (thread main, at /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/surfman-0.10.0/src/platform/unix/wayland/connection.rs:140)
fish: Job 1, './servo --screen-size=0x1' terminated by signal SIGSEGV (Address boundary error)
```
After:
```
╰─❯ ./servo --screen-size=0
Error: couldn't parse `0`: invalid resolution format
```
```
╰─❯ ./servo --screen-size=0x1
Error: couldn't parse `0x1`: width and height must be greater than 0
```

Taym95 commented on ports/servoshell/prefs.rs:
```diff
@@ -221,18 +221,43 @@ pub(crate) enum ArgumentParsingResult {
-            .collect::<Result<Vec<_>, std::num::ParseIntError>>()?;
-        Ok(Some(Size2D::new(components[0], components[1])))
+        let (width, height) = string
+            .split_once('x')
```
split_once('x') yields ("1", "1x1")?

Integral-Tech replied:
For screen resolution and window size, `split_once()` is sufficient.

- "1x1" -> ("1, "1")
-  "1x1x1" -> Error

Taym95 commented on ports/servoshell/prefs.rs:
```diff
@@ -221,18 +221,43 @@ pub(crate) enum ArgumentParsingResult {
-            .collect::<Result<Vec<_>, std::num::ParseIntError>>()?;
-        Ok(Some(Size2D::new(components[0], components[1])))
+        let (width, height) = string
+            .split_once('x')
```
Uppercase X not accepted?

Taym95 replied:
also whitespace handling?

Taym95 replied:
Inputs like "x1080" or "1920x" currently return ParseError (empty string parse) ?

Integral-Tech replied:
@Taym95 Resolved :)

Taym95 commented on ports/servoshell/prefs.rs:
```diff
@@ -221,18 +221,43 @@ pub(crate) enum ArgumentParsingResult {
+
 /// Parse a resolution string into a Size2D.
 fn parse_resolution_string(
     string: String,
```
not part of this PR but this needlessly takes ownership?

Integral-Tech replied:
@Taym95 `String` can't be replaced by `&str` here, as `bpaf` requires the trait bound `FromStr` for arguments.

```
error[E0277]: the trait bound `&str: FromStr` is not satisfied
```

Integral-Tech committed:
servoshell: Fix resolution string parsing

When passing an invalid resolution string (such as `1x1x1`) to the
`--screen-size` or `--window-size` argument, Servo starts without
any error. Additionally, if the width or height is set to 0, Servo
crashes with a SIGSEGV (Address boundary error).

This patch addresses the following issue by several changes:

1. Introduce a custom error type ParseResolutionError.
2. Replace the `split()` method with `split_once()`.
3. Make the capital 'X' an acceptable separator.
4. Add a check to prevent crashes when width or height is set to 0.

Signed-off-by: Integral <integral@member.fsf.org>

Integral-Tech commented:
@Taym95 If my pull request is acceptable, could you please approve the workflows?

yezhizhen reviewed (approved):
Looks reasonable!
